src/server/game/Entities/Object/ObjectGuid.h

```diff
@@ -233,18 +277,33 @@ class PackedGuid
         ByteBuffer _packedGuid;
 };
 
-template<HighGuid high>
-class ObjectGuidGenerator
+
+class ObjectGuidGeneratorBase
 {
-    public:
-        explicit ObjectGuidGenerator(uint32 start = 1) : _nextGuid(start) { }
+public:
+    ObjectGuidGeneratorBase(ObjectGuid::LowType start = 1) : _nextGuid(start) { }
 
-        void Set(uint32 val) { _nextGuid = val; }
-        uint32 Generate();
-        uint32 GetNextAfterMaxUsed() const { return _nextGuid; }
+    virtual void Set(uint32 val) { _nextGuid = val; }
+    virtual ObjectGuid::LowType Generate() = 0;
+    ObjectGuid::LowType GetNextAfterMaxUsed() const { return _nextGuid; }
 
-    private:
-        uint32 _nextGuid;
+protected:
+    static void HandleCounterOverflow(HighGuid high);
+    uint64 _nextGuid;
```

Comment:
When porting the entire codebase to use ObjectGuid instead of uint64 directly I defined ObjectGuid::LowType to not be convertible to/from any int type (to cause compiler errors), that should have take care of vast majority of them so I really would like to see the ones that were missed (with the exception of structures loaded from database)